File: go/noise/handshake.go
package noise

import (
	"errors"
	"fmt"
)

// Handshake pattern tokens
const (
	tokenE  = "e"  // ephemeral key
	tokenS  = "s"  // static key
	tokenEE = "ee" // DH(e, re)
	tokenES = "es" // DH(e, rs) or DH(s, re)
	tokenSE = "se" // DH(s, re) or DH(e, rs)
	tokenSS = "ss" // DH(s, rs)
)

// Pattern defines a Noise handshake pattern.
type Pattern struct {
	Name            string     // e.g., "IK"
	InitiatorPreMsg []string   // pre-message tokens for initiator (e.g., ["s"] for IK)
	ResponderPreMsg []string   // pre-message tokens for responder
	MessagePatterns [][]string // message patterns
}

// Predefined patterns
var (
	// PatternIK: Initiator knows responder's static key
	// <- s
	// ...
	// -> e, es, s, ss
	// <- e, ee, se
	PatternIK = Pattern{
		Name:            "IK",
		ResponderPreMsg: []string{tokenS},
		MessagePatterns: [][]string{
			{tokenE, tokenES, tokenS, tokenSS},
			{tokenE, tokenEE, tokenSE},
		},
	}

	// PatternXX: Mutual authentication, no prior knowledge
	// -> e
	// <- e, ee, s, es
	// -> s, se
	PatternXX = Pattern{
		Name: "XX",
		MessagePatterns: [][]string{
			{tokenE},
			{tokenE, tokenEE, tokenS, tokenES},
			{tokenS, tokenSE},
		},
	}

	// PatternNN: No authentication
	// -> e
	// <- e, ee
	PatternNN = Pattern{
		Name: "NN",
		MessagePatterns: [][]string{
			{tokenE},
			{tokenE, tokenEE},
		},
	}
)

// Config holds the configuration for a handshake.
type Config struct {
	Pattern      Pattern  // Handshake pattern
	Initiator    bool     // true if this side initiates
	LocalStatic  *KeyPair // Local static key pair (required for patterns with 's')
	RemoteStatic *Key     // Remote static public key (required for IK initiator)
	Prologue     []byte   // Optional prologue data
	PresharedKey *Key     // Optional PSK (for psk patterns)
}

// HandshakeState manages the state of a Noise handshake.
type HandshakeState struct {
	config Config
	ss     *SymmetricState

	localEphemeral  *KeyPair // Generated during handshake
	remoteEphemeral Key      // Received from peer
	remoteStatic    Key      // Received from peer (for XX pattern)

	msgIndex int  // Current message index
	finished bool // Handshake complete
}

// Errors
var (
	ErrHandshakeFinished   = errors.New("noise: handshake already finished")
	ErrHandshakeNotReady   = errors.New("noise: handshake not ready to split")
	ErrInvalidMessage      = errors.New("noise: invalid handshake message")
	ErrMissingLocalStatic  = errors.New("noise: missing local static key")
	ErrMissingRemoteStatic = errors.New("noise: missing remote static key")
)

// NewHandshakeState creates a new handshake state.
func NewHandshakeState(config Config) (*HandshakeState, error) {
	// Validate config
	if err := validateConfig(&config); err != nil {
		return nil, err
	}

	// Build protocol name
	protocolName := fmt.Sprintf("Noise_%s_25519_ChaChaPoly_BLAKE2s", config.Pattern.Name)

	hs := &HandshakeState{
		config: config,
		ss:     NewSymmetricState(protocolName),
	}

	// Mix in prologue
	hs.ss.MixHash(config.Prologue)

	// Process pre-messages
	if config.Initiator {
		// Initiator processes responder's pre-message (remote static)
		for _, token := range config.Pattern.ResponderPreMsg {
			if token == tokenS {
				if config.RemoteStatic == nil {
					return nil, ErrMissingRemoteStatic
				}
				hs.ss.MixHash(config.RemoteStatic[:])
				hs.remoteStatic = *config.RemoteStatic
			}
		}
		// Then initiator's pre-message
		for _, token := range config.Pattern.InitiatorPreMsg {
			if token == tokenS {
				hs.ss.MixHash(config.LocalStatic.Public[:])
			}
		}
	} else {
		// Responder processes initiator's pre-message (their static)
		for _, token := range config.Pattern.InitiatorPreMsg {
			if token == tokenS {
				if config.RemoteStatic == nil {
					return nil, ErrMissingRemoteStatic
				}
				hs.ss.MixHash(config.RemoteStatic[:])
				hs.remoteStatic = *config.RemoteStatic
			}
		}
		// Then responder's pre-message (our static)
		for _, token := range config.Pattern.ResponderPreMsg {
			if token == tokenS {
				hs.ss.MixHash(config.LocalStatic.Public[:])
			}
		}
	}

	return hs, nil
}

func validateConfig(config *Config) error {
	pattern := config.Pattern

	// Check if local static is needed
	needsLocalStatic := false
	for _, msg := range pattern.MessagePatterns {
		for _, token := range msg {
			if token == tokenS || token == tokenSS || token == tokenSE || token == tokenES {
				needsLocalStatic = true
				break
			}
		}
	}
	if needsLocalStatic && config.LocalStatic == nil {
		return ErrMissingLocalStatic
	}

	// For IK pattern, initiator needs remote static
	if pattern.Name == "IK" && config.Initiator && config.RemoteStatic == nil {
		return ErrMissingRemoteStatic
	}

	return nil
}

// WriteMessage generates the next handshake message.
// Returns the message to send to the peer.
func (hs *HandshakeState) WriteMessage(payload []byte) ([]byte, error) {
	if hs.finished {
		return nil, ErrHandshakeFinished
	}

	// Check if it's our turn
	myTurn := (hs.config.Initiator && hs.msgIndex%2 == 0) ||
		(!hs.config.Initiator && hs.msgIndex%2 == 1)
	if !myTurn {
		return nil, fmt.Errorf("noise: not our turn to write (msg %d)", hs.msgIndex)
	}

	if hs.msgIndex >= len(hs.config.Pattern.MessagePatterns) {
		return nil, ErrHandshakeFinished
	}

	tokens := hs.config.Pattern.MessagePatterns[hs.msgIndex]
	var msg []byte

	for _, token := range tokens {
		switch token {
		case tokenE:
			// Generate ephemeral key pair
			var err error
			hs.localEphemeral, err = GenerateKeyPair()
			if err != nil {
				return nil, fmt.Errorf("noise: failed to generate ephemeral key: %w", err)
			}
			msg = append(msg, hs.localEphemeral.Public[:]...)
			hs.ss.MixHash(hs.localEphemeral.Public[:])
			// If PSK mode, also mix key
			if hs.config.PresharedKey != nil {
				hs.ss.MixKey(hs.localEphemeral.Public[:])
			}

		case tokenS:
			// Encrypt and send static public key
			k := hs.ss.MixKey(nil) // Get encryption key
			encrypted := hs.ss.EncryptAndHash(&k, hs.config.LocalStatic.Public[:])
			msg = append(msg, encrypted...)

		case tokenEE:
			shared, err := hs.localEphemeral.DH(hs.remoteEphemeral)
			if err != nil {
				return nil, fmt.Errorf("noise: DH(e, re) failed: %w", err)
			}
			hs.ss.MixKey(shared[:])

		case tokenES:
			var shared Key
			var err error
			if hs.config.Initiator {
				// Initiator: DH(e, rs)
				shared, err = hs.localEphemeral.DH(hs.remoteStatic)
			} else {
				// Responder: DH(s, re)
				shared, err = hs.config.LocalStatic.DH(hs.remoteEphemeral)
			}
			if err != nil {
				return nil, fmt.Errorf("noise: DH(es) failed: %w", err)
			}
			hs.ss.MixKey(shared[:])

		case tokenSE:
			var shared Key
			var err error
			if hs.config.Initiator {
				// Initiator: DH(s, re)
				shared, err = hs.config.LocalStatic.DH(hs.remoteEphemeral)
			} else {
				// Responder: DH(e, rs)
				shared, err = hs.localEphemeral.DH(hs.remoteStatic)
			}
			if err != nil {
				return nil, fmt.Errorf("noise: DH(se) failed: %w", err)
			}
			hs.ss.MixKey(shared[:])

		case tokenSS:
			shared, err := hs.config.LocalStatic.DH(hs.remoteStatic)
			if err != nil {
				return nil, fmt.Errorf("noise: DH(s, rs) failed: %w", err)
			}
			hs.ss.MixKey(shared[:])
		}
	}

	// Encrypt payload if any
	if len(payload) > 0 || hs.msgIndex == len(hs.config.Pattern.MessagePatterns)-1 {
		// Get key for payload encryption
		k := hs.ss.MixKey(nil)
		encrypted := hs.ss.EncryptAndHash(&k, payload)
		msg = append(msg, encrypted...)
	}

	hs.msgIndex++
	if hs.msgIndex >= len(hs.config.Pattern.MessagePatterns) {
		hs.finished = true
	}

	return msg, nil
}

// ReadMessage processes a received handshake message.
// Returns the decrypted payload (if any).
func (hs *HandshakeState) ReadMessage(msg []byte) ([]byte, error) {
	if hs.finished {
		return nil, ErrHandshakeFinished
	}

	// Check if it's peer's turn
	myTurn := (hs.config.Initiator && hs.msgIndex%2 == 0) ||
		(!hs.config.Initiator && hs.msgIndex%2 == 1)
	if myTurn {
		return nil, fmt.Errorf("noise: not peer's turn to write (msg %d)", hs.msgIndex)
	}

	if hs.msgIndex >= len(hs.config.Pattern.MessagePatterns) {
		return nil, ErrHandshakeFinished
	}

	tokens := hs.config.Pattern.MessagePatterns[hs.msgIndex]
	offset := 0

	for _, token := range tokens {
		switch token {
		case tokenE:
			// Read remote ephemeral
			if offset+KeySize > len(msg) {
				return nil, ErrInvalidMessage
			}
			copy(hs.remoteEphemeral[:], msg[offset:offset+KeySize])
			offset += KeySize
			hs.ss.MixHash(hs.remoteEphemeral[:])
			if hs.config.PresharedKey != nil {
				hs.ss.MixKey(hs.remoteEphemeral[:])
			}

		case tokenS:
			// Decrypt remote static
			k := hs.ss.MixKey(nil)
			encryptedLen := KeySize + TagSize
			if offset+encryptedLen > len(msg) {
				return nil, ErrInvalidMessage
			}
			decrypted, err := hs.ss.DecryptAndHash(&k, msg[offset:offset+encryptedLen])
			if err != nil {
				return nil, fmt.Errorf("noise: failed to decrypt static key: %w", err)
			}
			copy(hs.remoteStatic[:], decrypted)
			offset += encryptedLen

		case tokenEE:
			shared, err := hs.localEphemeral.DH(hs.remoteEphemeral)
			if err != nil {
				return nil, fmt.Errorf("noise: DH(e, re) failed: %w", err)
			}
			hs.ss.MixKey(shared[:])

		case tokenES:
			var shared Key
			var err error
			if hs.config.Initiator {
				shared, err = hs.localEphemeral.DH(hs.remoteStatic)
			} else {
				shared, err = hs.config.LocalStatic.DH(hs.remoteEphemeral)
			}
			if err != nil {
				return nil, fmt.Errorf("noise: DH(es) failed: %w", err)
			}
			hs.ss.MixKey(shared[:])

		case tokenSE:
			var shared Key
			var err error
			if hs.config.Initiator {
				shared, err = hs.config.LocalStatic.DH(hs.remoteEphemeral)
			} else {
				shared, err = hs.localEphemeral.DH(hs.remoteStatic)
			}
			if err != nil {
				return nil, fmt.Errorf("noise: DH(se) failed: %w", err)
			}
			hs.ss.MixKey(shared[:])

		case tokenSS:
			shared, err := hs.config.LocalStatic.DH(hs.remoteStatic)
			if err != nil {
				return nil, fmt.Errorf("noise: DH(s, rs) failed: %w", err)
			}
			hs.ss.MixKey(shared[:])
		}
	}

	// Decrypt payload
	var payload []byte
	if offset < len(msg) {
		k := hs.ss.MixKey(nil)
		var err error
		payload, err = hs.ss.DecryptAndHash(&k, msg[offset:])
		if err != nil {
			return nil, fmt.Errorf("noise: failed to decrypt payload: %w", err)
		}
	}

	hs.msgIndex++
	if hs.msgIndex >= len(hs.config.Pattern.MessagePatterns) {
		hs.finished = true
	}

	return payload, nil
}

// IsFinished returns true if the handshake is complete.
func (hs *HandshakeState) IsFinished() bool {
	return hs.finished
}

// Split returns the transport CipherStates after handshake completion.
// Returns (send, recv) cipher states.
func (hs *HandshakeState) Split() (*CipherState, *CipherState, error) {
	if !hs.finished {
		return nil, nil, ErrHandshakeNotReady
	}

	cs1, cs2, err := hs.ss.Split()
	if err != nil {
		return nil, nil, err
	}

	if hs.config.Initiator {
		return cs1, cs2, nil
	}
	return cs2, cs1, nil
}

// RemoteStatic returns the remote peer's static public key.
// Only valid after the handshake message containing 's' has been processed.
func (hs *HandshakeState) RemoteStatic() Key {
	return hs.remoteStatic
}

// LocalEphemeral returns the local ephemeral public key.
// Only valid after WriteMessage has been called with 'e' token.
func (hs *HandshakeState) LocalEphemeral() Key {
	if hs.localEphemeral == nil {
		return Key{}
	}
	return hs.localEphemeral.Public
}

// Hash returns the current handshake hash.
// Can be used for channel binding after handshake.
func (hs *HandshakeState) Hash() [HashSize]byte {
	return hs.ss.Hash()
}

File: zig/src/noise/handshake.zig
//! Noise handshake patterns (IK, XX, NN).

const std = @import("std");
const mem = std.mem;

const keypair = @import("keypair.zig");
const cipher = @import("cipher.zig");
const state = @import("state.zig");
const c = @import("crypto.zig");

const Key = keypair.Key;
const KeyPair = keypair.KeyPair;
const CipherState = state.CipherState;
const SymmetricState = state.SymmetricState;
const key_size = keypair.key_size;
const tag_size = c.tag_size;
const hash_size = c.hash_size;

/// Handshake pattern.
pub const Pattern = enum {
    /// IK: Initiator knows responder's static key.
    IK,
    /// XX: Mutual authentication, no prior knowledge.
    XX,
    /// NN: No authentication.
    NN,

    fn name(self: Pattern) []const u8 {
        return switch (self) {
            .IK => "IK",
            .XX => "XX",
            .NN => "NN",
        };
    }

    fn responderPreMessage(self: Pattern) []const Token {
        return switch (self) {
            .IK => &[_]Token{.s},
            .XX, .NN => &[_]Token{},
        };
    }

    fn messagePatterns(self: Pattern) []const []const Token {
        return switch (self) {
            .IK => &[_][]const Token{
                &[_]Token{ .e, .es, .s, .ss },
                &[_]Token{ .e, .ee, .se },
            },
            .XX => &[_][]const Token{
                &[_]Token{.e},
                &[_]Token{ .e, .ee, .s, .es },
                &[_]Token{ .s, .se },
            },
            .NN => &[_][]const Token{
                &[_]Token{.e},
                &[_]Token{ .e, .ee },
            },
        };
    }
};

const Token = enum { e, s, ee, es, se, ss };

/// Handshake errors.
pub const Error = error{
    Finished,
    NotReady,
    InvalidMessage,
    MissingLocalStatic,
    MissingRemoteStatic,
    NotOurTurn,
    DecryptionFailed,
    DhFailed,
    LowOrderPoint,
};

/// Handshake configuration.
pub const Config = struct {
    pattern: Pattern,
    initiator: bool,
    local_static: ?KeyPair = null,
    remote_static: ?Key = null,
    prologue: []const u8 = "",
    preshared_key: ?Key = null,
};

/// Manages the state of a Noise handshake.
pub const HandshakeState = struct {
    pattern: Pattern,
    initiator: bool,
    local_static: ?KeyPair,
    remote_static: Key,
    preshared_key: ?Key,
    ss: SymmetricState,
    local_ephemeral: ?KeyPair,
    remote_ephemeral: Key,
    msg_index: usize,
    finished: bool,

    /// Creates a new handshake state.
    pub fn init(config: Config) Error!HandshakeState {
        // Validate config
        try validateConfig(config);

        // Build protocol name
        var protocol_buf: [64]u8 = undefined;
        const protocol_name = std.fmt.bufPrint(&protocol_buf, "Noise_{s}_25519_ChaChaPoly_BLAKE2s", .{config.pattern.name()}) catch unreachable;

        var ss = SymmetricState.init(protocol_name);

        // Mix prologue
        ss.mixHash(config.prologue);

        var remote_static = Key.zero;

        // Process pre-messages
        if (config.initiator) {
            for (config.pattern.responderPreMessage()) |token| {
                if (token == .s) {
                    const rs = config.remote_static orelse return Error.MissingRemoteStatic;
                    ss.mixHash(rs.asBytes());
                    remote_static = rs;
                }
            }
        } else {
            for (config.pattern.responderPreMessage()) |token| {
                if (token == .s) {
                    const ls = config.local_static orelse return Error.MissingLocalStatic;
                    ss.mixHash(ls.public.asBytes());
                }
            }
        }

        return .{
            .pattern = config.pattern,
            .initiator = config.initiator,
            .local_static = config.local_static,
            .remote_static = remote_static,
            .preshared_key = config.preshared_key,
            .ss = ss,
            .local_ephemeral = null,
            .remote_ephemeral = Key.zero,
            .msg_index = 0,
            .finished = false,
        };
    }

    fn validateConfig(config: Config) Error!void {
        const needs_local_static = config.pattern == .IK or config.pattern == .XX;
        if (needs_local_static and config.local_static == null) {
            return Error.MissingLocalStatic;
        }
        if (config.pattern == .IK and config.initiator and config.remote_static == null) {
            return Error.MissingRemoteStatic;
        }
    }

    /// Generates the next handshake message.
    pub fn writeMessage(self: *HandshakeState, payload: []const u8, out: []u8) Error!usize {
        if (self.finished) return Error.Finished;

        const my_turn = (self.initiator and self.msg_index % 2 == 0) or
            (!self.initiator and self.msg_index % 2 == 1);
        if (!my_turn) return Error.NotOurTurn;

        const patterns = self.pattern.messagePatterns();
        if (self.msg_index >= patterns.len) return Error.Finished;

        const tokens = patterns[self.msg_index];
        var offset: usize = 0;

        for (tokens) |token| {
            switch (token) {
                .e => {
                    const ephemeral = KeyPair.generate();
                    @memcpy(out[offset..][0..key_size], ephemeral.public.asBytes());
                    offset += key_size;
                    self.ss.mixHash(ephemeral.public.asBytes());
                    if (self.preshared_key != null) {
                        _ = self.ss.mixKey(ephemeral.public.asBytes());
                    }
                    self.local_ephemeral = ephemeral;
                },
                .s => {
                    const ls = self.local_static orelse return Error.MissingLocalStatic;
                    const k = self.ss.mixKey("");
                    self.ss.encryptAndHash(&k, ls.public.asBytes(), out[offset..][0 .. key_size + tag_size]);
                    offset += key_size + tag_size;
                },
                .ee => {
                    const le = self.local_ephemeral orelse return Error.InvalidMessage;
                    const shared = le.dh(self.remote_ephemeral) catch return Error.DhFailed;
                    _ = self.ss.mixKey(shared.asBytes());
                },
                .es => {
                    const shared = if (self.initiator) blk: {
                        const le = self.local_ephemeral orelse return Error.InvalidMessage;
                        break :blk le.dh(self.remote_static) catch return Error.DhFailed;
                    } else blk: {
                        const ls = self.local_static orelse return Error.MissingLocalStatic;
                        break :blk ls.dh(self.remote_ephemeral) catch return Error.DhFailed;
                    };
                    _ = self.ss.mixKey(shared.asBytes());
                },
                .se => {
                    const shared = if (self.initiator) blk: {
                        const ls = self.local_static orelse return Error.MissingLocalStatic;
                        break :blk ls.dh(self.remote_ephemeral) catch return Error.DhFailed;
                    } else blk: {
                        const le = self.local_ephemeral orelse return Error.InvalidMessage;
                        break :blk le.dh(self.remote_static) catch return Error.DhFailed;
                    };
                    _ = self.ss.mixKey(shared.asBytes());
                },
                .ss => {
                    const ls = self.local_static orelse return Error.MissingLocalStatic;
                    const shared = ls.dh(self.remote_static) catch return Error.DhFailed;
                    _ = self.ss.mixKey(shared.asBytes());
                },
            }
        }

        // Encrypt payload
        if (payload.len > 0 or self.msg_index == patterns.len - 1) {
            const k = self.ss.mixKey("");
            self.ss.encryptAndHash(&k, payload, out[offset..][0 .. payload.len + tag_size]);
            offset += payload.len + tag_size;
        }

        self.msg_index += 1;
        if (self.msg_index >= patterns.len) {
            self.finished = true;
        }

        return offset;
    }

    /// Processes a received handshake message.
    pub fn readMessage(self: *HandshakeState, msg: []const u8, payload_out: []u8) Error!usize {
        if (self.finished) return Error.Finished;

        const my_turn = (self.initiator and self.msg_index % 2 == 0) or
            (!self.initiator and self.msg_index % 2 == 1);
        if (my_turn) return Error.NotOurTurn;

        const patterns = self.pattern.messagePatterns();
        if (self.msg_index >= patterns.len) return Error.Finished;

        const tokens = patterns[self.msg_index];
        var offset: usize = 0;

        for (tokens) |token| {
            switch (token) {
                .e => {
                    if (offset + key_size > msg.len) return Error.InvalidMessage;
                    self.remote_ephemeral = Key.fromSlice(msg[offset..][0..key_size]) catch return Error.InvalidMessage;
                    offset += key_size;
                    self.ss.mixHash(self.remote_ephemeral.asBytes());
                    if (self.preshared_key != null) {
                        _ = self.ss.mixKey(self.remote_ephemeral.asBytes());
                    }
                },
                .s => {
                    const encrypted_len = key_size + tag_size;
                    if (offset + encrypted_len > msg.len) return Error.InvalidMessage;
                    const k = self.ss.mixKey("");
                    var rs_bytes: [key_size]u8 = undefined;
                    self.ss.decryptAndHash(&k, msg[offset..][0..encrypted_len], &rs_bytes) catch return Error.DecryptionFailed;
                    self.remote_static = Key.fromBytes(rs_bytes);
                    offset += encrypted_len;
                },
                .ee => {
                    const le = self.local_ephemeral orelse return Error.InvalidMessage;
                    const shared = le.dh(self.remote_ephemeral) catch return Error.DhFailed;
                    _ = self.ss.mixKey(shared.asBytes());
                },
                .es => {
                    const shared = if (self.initiator) blk: {
                        const le = self.local_ephemeral orelse return Error.InvalidMessage;
                        break :blk le.dh(self.remote_static) catch return Error.DhFailed;
                    } else blk: {
                        const ls = self.local_static orelse return Error.MissingLocalStatic;
                        break :blk ls.dh(self.remote_ephemeral) catch return Error.DhFailed;
                    };
                    _ = self.ss.mixKey(shared.asBytes());
                },
                .se => {
                    const shared = if (self.initiator) blk: {
                        const ls = self.local_static orelse return Error.MissingLocalStatic;
                        break :blk ls.dh(self.remote_ephemeral) catch return Error.DhFailed;
                    } else blk: {
                        const le = self.local_ephemeral orelse return Error.InvalidMessage;
                        break :blk le.dh(self.remote_static) catch return Error.DhFailed;
                    };
                    _ = self.ss.mixKey(shared.asBytes());
                },
                .ss => {
                    const ls = self.local_static orelse return Error.MissingLocalStatic;
                    const shared = ls.dh(self.remote_static) catch return Error.DhFailed;
                    _ = self.ss.mixKey(shared.asBytes());
                },
            }
        }

        // Decrypt payload
        var payload_len: usize = 0;
        if (offset < msg.len) {
            const k = self.ss.mixKey("");
            payload_len = msg.len - offset - tag_size;
            self.ss.decryptAndHash(&k, msg[offset..], payload_out[0..payload_len]) catch return Error.DecryptionFailed;
        }

        self.msg_index += 1;
        if (self.msg_index >= self.pattern.messagePatterns().len) {
            self.finished = true;
        }

        return payload_len;
    }

    /// Returns true if handshake is complete.
    pub fn isFinished(self: HandshakeState) bool {
        return self.finished;
    }

    /// Splits into transport CipherStates.
    pub fn split(self: *const HandshakeState) Error!struct { CipherState, CipherState } {
        if (!self.finished) return Error.NotReady;

        const cs1, const cs2 = self.ss.split();
        if (self.initiator) {
            return .{ cs1, cs2 };
        } else {
            return .{ cs2, cs1 };
        }
    }

    /// Returns the remote static public key.
    pub fn getRemoteStatic(self: HandshakeState) Key {
        return self.remote_static;
    }

    /// Returns the handshake hash.
    pub fn getHash(self: *const HandshakeState) *const [hash_size]u8 {
        return self.ss.getHash();
    }
};

// Tests
test "handshake IK" {
    const initiator_static = KeyPair.generate();
    const responder_static = KeyPair.generate();

    var initiator = try HandshakeState.init(.{
        .pattern = .IK,
        .initiator = true,
        .local_static = initiator_static,
        .remote_static = responder_static.public,
    });

    var responder = try HandshakeState.init(.{
        .pattern = .IK,
        .initiator = false,
        .local_static = responder_static,
    });

    // Message 1
    var msg1: [256]u8 = undefined;
    const msg1_len = try initiator.writeMessage("", &msg1);

    var payload1: [64]u8 = undefined;
    _ = try responder.readMessage(msg1[0..msg1_len], &payload1);
    try std.testing.expect(responder.getRemoteStatic().eql(initiator_static.public));

    // Message 2
    var msg2: [256]u8 = undefined;
    const msg2_len = try responder.writeMessage("", &msg2);

    var payload2: [64]u8 = undefined;
    _ = try initiator.readMessage(msg2[0..msg2_len], &payload2);

    try std.testing.expect(initiator.isFinished());
    try std.testing.expect(responder.isFinished());

    // Test transport
    var send_i, var recv_i = try initiator.split();
    var send_r, var recv_r = try responder.split();

    const plaintext = "hello from initiator";
    var ct: [plaintext.len + tag_size]u8 = undefined;
    send_i.encrypt(plaintext, "", &ct);

    var pt: [plaintext.len]u8 = undefined;
    try recv_r.decrypt(&ct, "", &pt);
    try std.testing.expectEqualSlices(u8, plaintext, &pt);

    const reply = "hello from responder";
    var ct2: [reply.len + tag_size]u8 = undefined;
    send_r.encrypt(reply, "", &ct2);

    var pt2: [reply.len]u8 = undefined;
    try recv_i.decrypt(&ct2, "", &pt2);
    try std.testing.expectEqualSlices(u8, reply, &pt2);
}

test "handshake XX" {
    const initiator_static = KeyPair.generate();
    const responder_static = KeyPair.generate();

    var initiator = try HandshakeState.init(.{
        .pattern = .XX,
        .initiator = true,
        .local_static = initiator_static,
    });

    var responder = try HandshakeState.init(.{
        .pattern = .XX,
        .initiator = false,
        .local_static = responder_static,
    });

    // Message 1: -> e
    var msg1: [256]u8 = undefined;
    const msg1_len = try initiator.writeMessage("", &msg1);
    var p1: [64]u8 = undefined;
    _ = try responder.readMessage(msg1[0..msg1_len], &p1);

    // Message 2: <- e, ee, s, es
    var msg2: [256]u8 = undefined;
    const msg2_len = try responder.writeMessage("", &msg2);
    var p2: [64]u8 = undefined;
    _ = try initiator.readMessage(msg2[0..msg2_len], &p2);
    try std.testing.expect(initiator.getRemoteStatic().eql(responder_static.public));

    // Message 3: -> s, se
    var msg3: [256]u8 = undefined;
    const msg3_len = try initiator.writeMessage("", &msg3);
    var p3: [64]u8 = undefined;
    _ = try responder.readMessage(msg3[0..msg3_len], &p3);
    try std.testing.expect(responder.getRemoteStatic().eql(initiator_static.public));

    try std.testing.expect(initiator.isFinished());
    try std.testing.expect(responder.isFinished());
}

test "handshake NN" {
    var initiator = try HandshakeState.init(.{
        .pattern = .NN,
        .initiator = true,
    });

    var responder = try HandshakeState.init(.{
        .pattern = .NN,
        .initiator = false,
    });

    // Message 1: -> e
    var msg1: [256]u8 = undefined;
    const msg1_len = try initiator.writeMessage("", &msg1);
    var p1: [64]u8 = undefined;
    _ = try responder.readMessage(msg1[0..msg1_len], &p1);

    // Message 2: <- e, ee
    var msg2: [256]u8 = undefined;
    const msg2_len = try responder.writeMessage("", &msg2);
    var p2: [64]u8 = undefined;
    _ = try initiator.readMessage(msg2[0..msg2_len], &p2);

    try std.testing.expect(initiator.isFinished());
    try std.testing.expect(responder.isFinished());

    var send_i, _ = try initiator.split();
    _, var recv_r = try responder.split();

    const plaintext = "NN test";
    var ct: [plaintext.len + tag_size]u8 = undefined;
    send_i.encrypt(plaintext, "", &ct);

    var pt: [plaintext.len]u8 = undefined;
    try recv_r.decrypt(&ct, "", &pt);
    try std.testing.expectEqualSlices(u8, plaintext, &pt);
}

test "handshake hash" {
    const initiator_static = KeyPair.generate();
    const responder_static = KeyPair.generate();

    var initiator = try HandshakeState.init(.{
        .pattern = .IK,
        .initiator = true,
        .local_static = initiator_static,
        .remote_static = responder_static.public,
    });

    var responder = try HandshakeState.init(.{
        .pattern = .IK,
        .initiator = false,
        .local_static = responder_static,
    });

    var msg1: [256]u8 = undefined;
    const msg1_len = try initiator.writeMessage("", &msg1);
    var p1: [64]u8 = undefined;
    _ = try responder.readMessage(msg1[0..msg1_len], &p1);

    var msg2: [256]u8 = undefined;
    const msg2_len = try responder.writeMessage("", &msg2);
    var p2: [64]u8 = undefined;
    _ = try initiator.readMessage(msg2[0..msg2_len], &p2);

    try std.testing.expectEqualSlices(u8, initiator.getHash(), responder.getHash());
}

test "handshake errors" {
    // Missing local static for IK
    const rs = KeyPair.generate();
    try std.testing.expectError(Error.MissingLocalStatic, HandshakeState.init(.{
        .pattern = .IK,
        .initiator = true,
        .remote_static = rs.public,
    }));

    // Missing remote static for IK initiator
    const ls = KeyPair.generate();
    try std.testing.expectError(Error.MissingRemoteStatic, HandshakeState.init(.{
        .pattern = .IK,
        .initiator = true,
        .local_static = ls,
    }));
}

test "handshake wrong turn" {
    var initiator = try HandshakeState.init(.{
        .pattern = .NN,
        .initiator = true,
    });

    var responder = try HandshakeState.init(.{
        .pattern = .NN,
        .initiator = false,
    });

    // Responder tries to write first
    var buf: [256]u8 = undefined;
    try std.testing.expectError(Error.NotOurTurn, responder.writeMessage("", &buf));

    // Initiator tries to read first
    var p: [64]u8 = undefined;
    try std.testing.expectError(Error.NotOurTurn, initiator.readMessage("", &p));
}

test "split before finish" {
    const initiator = try HandshakeState.init(.{
        .pattern = .NN,
        .initiator = true,
    });

    try std.testing.expectError(Error.NotReady, initiator.split());
}
